src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntityResourceProcessor.cs

```diff
@@ -0,0 +1,483 @@
+    private readonly List<EntityResourceEntryDefinition> _resourceEntryDefinitionsToAdd = new();
+    private readonly List<EntityResourceTotalsHistory> _resourceTotalsHistoryToAdd = new();
+    private readonly List<EntityResourceBalanceHistory> _resourceBalanceHistoryToAdd = new();
+    private readonly List<EntityResourceVaultEntryDefinition> _resourceVaultEntryDefinitionsToAdd = new();
+    private readonly List<EntityResourceVaultTotalsHistory> _resourceVaultTotalsHistoryToAdd = new();
+
+    public EntityResourceProcessor(ProcessorContext context)
+    {
+        _context = context;
+    }
+
+    public void VisitUpsert(CoreModel.Substate substateData, ReferencedEntity referencedEntity, long stateVersion, CoreModel.IUpsertedSubstate substate)
+    {
+        if (substateData is CoreModel.FungibleVaultFieldBalanceSubstate fungibleBalanceSubstate)
+        {
+            var vaultEntity = referencedEntity.GetDatabaseEntity<InternalFungibleVaultEntity>();
+            var amount = TokenAmount.FromDecimalString(fungibleBalanceSubstate.Value.Amount);
+            var previousAmountRaw = (substate.PreviousValue?.SubstateData as CoreModel.FungibleVaultFieldBalanceSubstate)?.Value.Amount;
+            var previousAmount = previousAmountRaw == null ? TokenAmount.Zero : TokenAmount.FromDecimalString(previousAmountRaw);
+            var delta = amount - previousAmount;
+
+            if (!vaultEntity.IsRoyaltyVault)
+            {
+                _observedVaultChanges.Add(new VaultChange(ResourceType.Fungible, vaultEntity, delta, stateVersion));
+            }
+        }
+
+        if (substateData is CoreModel.NonFungibleVaultFieldBalanceSubstate nonFungibleBalanceSubstate)
+        {
+            var vaultEntity = referencedEntity.GetDatabaseEntity<InternalNonFungibleVaultEntity>();
+            var amount = TokenAmount.FromDecimalString(nonFungibleBalanceSubstate.Value.Amount);
+            var previousAmountRaw = (substate.PreviousValue?.SubstateData as CoreModel.NonFungibleVaultFieldBalanceSubstate)?.Value.Amount;
+            var previousAmount = previousAmountRaw == null ? TokenAmount.Zero : TokenAmount.FromDecimalString(previousAmountRaw);
+            var delta = amount - previousAmount;
+
+            _observedVaultChanges.Add(new VaultChange(ResourceType.NonFungible, vaultEntity, delta, stateVersion));
+        }
+    }
+
+    public async Task LoadDependencies()
+    {
+        _existingResourceDefinitions.AddRange(await ExistingEntityResourceEntryDefinitions());
+        _mostRecentResourceTotalsHistory.AddRange(await MostRecentEntityResourceTotalsHistory());
+        _mostRecentResourceBalanceHistory.AddRange(await MostRecentResourceBalanceHistory());
+        _existingResourceVaultDefinitions.AddRange(await ExistingEntityResourceVaultEntryDefinitions());
+        _mostRecentResourceVaultTotalsHistory.AddRange(await MostRecentEntityResourceVaultTotalsHistory());
+    }
+
+    public void ProcessChanges()
```

Comment:
Do we need to process all these changes in one "processor"? Is it possible to extract smaller ones? Or at least divide that logic into separate methods?

Reply:
Yeah, it seems to be the case. I guess those foreach loop bodies could have been split into separate method though.